File: code/ch20/20.1.2. 73.setZeroes.py
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """

        def setRowZeros(matrix: List[List[int]], i: int) -> None:
            C = len(matrix[0])
            matrix[i] = [0] * C

        def setColZeros(matrix: List[List[int]], j: int) -> None:
            R = len(matrix)
            for i in range(R):
                matrix[i][j] = 0

        isCol = False
        R = len(matrix)
        C = len(matrix[0])

        for i in range(R):
            if matrix[i][0] == 0:
                isCol = True
            for j in range(1, C):
                if matrix[i][j] == 0:
                    matrix[i][0] = 0
                    matrix[0][j] = 0
        for j in range(1, C):
            if matrix[0][j] == 0:
                setColZeros(matrix, j)

        for i in range(R):
            if matrix[i][0] == 0:
                setRowZeros(matrix, i)

        if isCol:
            setColZeros(matrix, 0)

File: code/ch20/20.1.2. 73.setZeroes.cpp
#include <vector>
using namespace std;

class Solution {
public:
  void setRowZeros(vector<vector<int> >& matrix, int i) {
    for (int j = 0; j < matrix[i].size(); j++)
      matrix[i][j] = 0;
  }
  void setColZeros(vector<vector<int> >& matrix, int j) {
    for (int i = 0; i < matrix.size(); i++)
      matrix[i][j] = 0;
  }
  void setZeroes(vector<vector<int> >& matrix) {
    bool isCol = false;
    int R = matrix.size(), C = matrix[0].size();
    for (int i = 0; i < R; i++) {
      if (matrix[i][0] == 0) isCol = true;
      for (int j = 1; j < C; j++)
        if (matrix[i][j] == 0) {
          matrix[i][0] = 0;
          matrix[0][j] = 0;
        }
    }
    for (int j = 1; j < C; j++)
      if (matrix[0][j] == 0)
        setColZeros(matrix, j);
    for (int i = 0; i < R; i++)
      if (matrix[i][0] == 0)
        setRowZeros(matrix, i);
    if (isCol) setColZeros(matrix, 0);
  }
};
